### PR TITLE
Implement read_volatile and write_volatile

### PR DESCRIPTION
**Pull request by Amanieu**
Tracking issue: #31756
RFC: rust-lang/rfcs#1467

I've made these unstable for now. Should they be stabilized straight away since we've had plenty of experience with people using the unstable intrinsics?

### CODE REVIEW

File: src/libcore/ptr.rs [hunk 1]
```diff
@@ -161,6 +161,54 @@ pub unsafe fn write<T>(dst: *mut T, src: T) {
     intrinsics::move_val_init(&mut *dst, src)
 }
 
+/// Performs a volatile read of the value from `src` without moving it. This
+/// leaves the memory in `src` unchanged.
+///
+/// Volatile operations are intended to act on I/O memory, and are guaranteed
+/// to not be elided or reordered by the compiler across other volatile
+/// operations. See the LLVM documentation on [[volatile]].
```
Comment (alexcrichton) [inline src/libcore/ptr.rs]: Oh weird, does markdown work like this? I thought only one set of brackets was needed?

**Comment by alexcrichton** [comment]
Looks good to me, thanks @Amanieu! Let's land these as unstable yeah to start out and we can likely stabilize them during the next cycle!

File: src/libcore/ptr.rs (hunk 1 above)
Comment (Amanieu) [inline src/libcore/ptr.rs]: I just copied this from the documentation in the `intrinsics` module.

Comment (alexcrichton) [inline src/libcore/ptr.rs]: Ah, ok!

**Comment by alexcrichton** [comment]
Oh right, sorry, forgot to ask as well, but can you add some tests which just exercise these a bit?

**Comment by Amanieu** [comment]
Is there a way to make a test that checks that some specific LLVM IR is generated?

**Comment by alexcrichton** [comment]
In `src/test/codegen`, yes, but I'd also be fine with just a smoke test that *uses* these to make sure the compiler doesn't panic on them or something weird like that

**Comment by Amanieu** [comment]
I added a test.

**Pull request merged**